lib/wasi/src/runtime/module_cache/types.rs

```diff
@@ -0,0 +1,164 @@
+use std::{
+    fmt::{self, Debug, Display, Formatter},
+    ops::Deref,
+    path::PathBuf,
+};
+
+use sha2::{Digest, Sha256};
+use wasmer::{Engine, Module};
+
+use crate::runtime::module_cache::AndThen;
+
+/// A cache for compiled WebAssembly modules.
+///
+/// ## Assumptions
+///
+/// Implementations can assume that cache keys are unique and that using the
+/// same key to load or save will always result in the "same" module.
+///
+/// Implementations can also assume that [`ModuleCache::load()`] will
+/// be called more often than [`ModuleCache::save()`] and optimise
+/// their caching strategy accordingly.
+#[async_trait::async_trait]
+pub trait ModuleCache: Debug {
+    async fn load(&self, key: Key, engine: &Engine) -> Result<Module, CacheError>;
+
+    async fn save(&self, key: Key, module: &Module) -> Result<(), CacheError>;
```

Comment:
Thinking more about this, probably this should return a `SaveCacheError` (enum with states: WriteError(wasmer::SerializeError), Generic(std::error::Error))

Reply:
I'd prefer to leave this as a single error for now. Let's wait until the `ModuleCache` trait gets some real-world use before doing more tweaks.